Mysql compatibility

maxmarcon opened the pull request:
Here's the PR for issue #52 

Looking forward to your comments.

kelvinst commented on mix.exs:
```diff
@@ -50,7 +50,8 @@ defmodule Triplex.Mixfile do
       {:ecto, "~> 2.1"},
       {:postgrex, ">= 0.11.0"},
-
+      {:mariaex, "~> 0.8.2"},
```
So, let's make this one and `:postgrex` dependencies optional, since the user will probably choose between one and another.

maxmarcon replied:
I kept only mariaex optional, because otherwise your test suite fails. It's weird, I would have expected the "optional" requirement to have an effect only if the project is included by another one: https://hexdocs.pm/mix/Mix.Tasks.Deps.html but I must be missing something.

kelvinst commented on lib/triplex.ex:
```diff
@@ -138,12 +151,23 @@ defmodule Triplex do
+          case SQL.query(repo, "DELETE FROM #{Triplex.config().tenant_table} WHERE NAME = ?", [tenant]) do
+            {:error, msg} -> {:error, msg}
+            _ -> {:ok, tenant}
+          end
```
We could refactor this nested `case`s to use a `with`

maxmarcon replied:
Done

kelvinst commented on lib/triplex.ex:
```diff
@@ -157,18 +181,23 @@ defmodule Triplex do
       {:error, reserved_message(new_tenant)}
-    else
+    repo.__adapter__ == Ecto.Adapters.MySQL ->
+      {:error, "you cannot rename tenants in a MySQL database. You will have to drop the tenant and recreate it with a different name"}
```
Well, to drop the tenant and create another is not a good option, since the user will lose all his tenant data. Let's remove this part and stay only with: `"you cannot rename tenants in a MySQL database"`

maxmarcon replied:
Good point, changed.

kelvinst commented on lib/triplex.ex:
```diff
@@ -177,12 +206,16 @@ defmodule Triplex do
+    sql = case repo.__adapter__ do
+      Ecto.Adapters.MySQL ->
+        "SELECT name FROM #{config().tenant_table}"
+      _ ->
```
Let's "let it fail" and put `Ecto.Adapters.Postgres` on this case clause ;)

maxmarcon replied:
Yup

kelvinst commented on lib/triplex.ex:
```diff
@@ -197,14 +230,19 @@ defmodule Triplex do
+      sql = case repo.__adapter__ do
+        Ecto.Adapters.MySQL ->
+          "SELECT COUNT(*) FROM #{config().tenant_table} WHERE name = ?"
+        _ ->
```
Let's "let it fail" and put `Ecto.Adapters.Postgres` on this case clause ;)

maxmarcon replied:
Yup

maxmarcon commented:
@kelvinst thanks for your comments. I addressed them the best I could.

I also did 2 additional things:

1) I adapted the tests to work under MySQL. There's some ugliness caused by DDL operations in MySQL not being compatible with nested transactions. I documented it in `triplex_test.exs`.

2) I modified the `triplex.migrations` task to list the state of the migrations for each tenant. So far, it wasn't really working because it was only using `Ecto.Migrator.migrations`, which gives you the list of migrations with their status but is not "tenant aware" (you can't pass it a prefix). Therefore, it was always showing the migrations as being down. I changed it to combine that information with the result of `Ecto.Migrator.migrated_versions`, which accepts a prefix and you can use to obtain the real status of the migrations in a tenant.

kelvinst commented:
Hey @maxmarcon cool! Sorry for the delay, new job. 😬

I'll review it soon, I promise. 😄

kelvinst commented on lib/mix/tasks/triplex.init.tenant.ex:
```diff
@@ -0,0 +1,58 @@
+defmodule Mix.Tasks.Triplex.Init.Tenant do
```
What about renaming this task to `triplex.mysql.install`?

maxmarcon replied:
sure thing
